internal/app/issue/issue_handler.go

```diff
@@ -0,0 +1,302 @@
+	}
+
+	is.eventRegistry.PushEvent(&ListIssuesEvent{
+		Filter:  filter,
+		Options: options,
+		Issues:  &issueList,
+	})
+
+	return &issueList, nil
+}
+
+func (is *issueHandler) CreateIssue(issue *entity.Issue) (*entity.Issue, error) {
+	f := &entity.IssueFilter{
+		PrimaryName: []*string{&issue.PrimaryName},
+	}
+
+	l := logrus.WithFields(logrus.Fields{
+		"event":  CreateIssueEventName,
+		"object": issue,
+		"filter": f,
+	})
+
+	issues, err := is.ListIssues(f, &entity.IssueListOptions{})
+
+	if err != nil {
+		l.Error(err)
+		return nil, NewIssueHandlerError("Internal error while creating issue.")
+	}
+
+	if len(issues.Elements) > 0 {
+		return nil, NewIssueHandlerError(fmt.Sprintf("Duplicated entry %s for primaryName.", issue.PrimaryName))
+	}
+
+	newIssue, err := is.database.CreateIssue(issue)
+
+	if err != nil {
+		l.Error(err)
+		return nil, NewIssueHandlerError("Internal error while creating issue.")
+	}
+
+	is.eventRegistry.PushEvent(&CreateIssueEvent{Issue: newIssue})
+	return newIssue, nil
+}
+
+func (is *issueHandler) UpdateIssue(issue *entity.Issue) (*entity.Issue, error) {
+	l := logrus.WithFields(logrus.Fields{
+		"event":  UpdateIssueEventName,
+		"object": issue,
+	})
+
+	err := is.database.UpdateIssue(issue)
+
+	if err != nil {
+		l.Error(err)
+		return nil, NewIssueHandlerError("Internal error while updating issue.")
+	}
+
+	issueResult, err := is.ListIssues(&entity.IssueFilter{Id: []*int64{&issue.Id}}, &entity.IssueListOptions{})
+
+	if err != nil {
+		l.Error(err)
+		return nil, NewIssueHandlerError("Internal error while retrieving updated issue.")
+	}
+
+	if len(issueResult.Elements) != 1 {
+		l.Error(err)
+		return nil, NewIssueHandlerError("Multiple issues found.")
+	}
+
+	is.eventRegistry.PushEvent(&UpdateIssueEvent{Issue: issue})
+	return issueResult.Elements[0].Issue, nil
+}
+
+func (is *issueHandler) DeleteIssue(id int64) error {
+	l := logrus.WithFields(logrus.Fields{
+		"event": DeleteIssueEventName,
+		"id":    id,
+	})
+
+	err := is.database.DeleteIssue(id)
+
+	if err != nil {
+		l.Error(err)
+		return NewIssueHandlerError("Internal error while deleting issue.")
+	}
+
+	is.eventRegistry.PushEvent(&DeleteIssueEvent{IssueID: id})
+	return nil
+}
+
+func (is *issueHandler) AddComponentVersionToIssue(issueId, componentVersionId int64) (*entity.Issue, error) {
+	l := logrus.WithFields(logrus.Fields{
+		"event": AddComponentVersionToIssueEventName,
+		"id":    issueId,
+	})
+
+	err := is.database.AddComponentVersionToIssue(issueId, componentVersionId)
+
+	if err != nil {
+		l.Error(err)
+		return nil, NewIssueHandlerError("Internal error while adding component version to issue.")
+	}
+
+	is.eventRegistry.PushEvent(&AddComponentVersionToIssueEvent{
+		IssueID:            issueId,
+		ComponentVersionID: componentVersionId,
+	})
+
+	return is.GetIssue(issueId)
+}
+
+func (is *issueHandler) RemoveComponentVersionFromIssue(issueId, componentVersionId int64) (*entity.Issue, error) {
+	l := logrus.WithFields(logrus.Fields{
```

Comment:
I would not create logger instance in every call of handler, I think it could be nice to have logger interface used in each handler and this logger interface could be component of the handler. It could be also advantage that handlers would get rid of logrus dependency.
Furthermore it is mostly used for error handling, so in typical handler call it is not used at all. Maybe we could add extra factory function to create logger only when it is needed.

Reply:
Makes sense

Reply:
I think a factory may not be necessary, but I see your point that it's pointless to create the logger with context and then later not use it.